modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/IgniteIndexNamesTest.java

```diff
@@ -0,0 +1,206 @@
+import org.junit.runners.JUnit4;
+
+/**
+ * Testing names of indexes.
+ */
+@RunWith(JUnit4.class)
+public class IgniteIndexNamesTest extends GridCommonAbstractTest {
+    /** */
+    private static final String TABLE_CACHE_NAME = "FOO";
+
+    /** */
+    private static final int NODES_COUNT = 2;
+
+    /** */
+    private boolean useJdbc;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setDataStorageConfiguration(
+            new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(
+                    new DataRegionConfiguration()
+                        .setPersistenceEnabled(true)
+                )
+        );
+
+        return cfg;
+    }
+
+    /** */
+    @Before
+    public void before() throws Exception {
+        // Being paranoid.
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** */
+    @After
+    public void after() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** */
+    @Test
+    public void testNamesUseJava() throws Exception {
+        useJdbc = false;
+
+        check();
+    }
+
+    /** */
+    @Test
+    public void testNamesUseJdbc() throws Exception {
+        useJdbc = true;
+
+        check();
+    }
+
+    /** */
+    private void check() throws Exception {
+        startGrids(NODES_COUNT);
+
+        grid(0).cluster().active(true);
+
+        // Create gateway cache.
+        grid(0).createCache(DEFAULT_CACHE_NAME);
+
+        runSql("CREATE TABLE PUBLIC.FOO (A INT, B INT, C INT PRIMARY KEY) WITH \"CACHE_NAME=FOO\"");
+
+        List<String> createIndexQueries = Arrays.asList(
+            "CREATE INDEX ON PUBLIC.FOO (a)",
+            "CREATE INDEX ON PUBLIC.FOO (b)",
+            "CREATE INDEX ON PUBLIC.FOO (c)",
+            "CREATE INDEX \"idx\" ON PUBLIC.FOO (a)",
+            "CREATE INDEX \"IDX\" ON PUBLIC.FOO (a)",
+            "CREATE INDEX \"iDx\" ON PUBLIC.FOO (a)",
+            "CREATE INDEX \"\" ON PUBLIC.FOO (a)",
+            "CREATE INDEX \" \" ON PUBLIC.FOO (a)"
+        );
+
+        for (String sql : createIndexQueries)
+            runSql(sql);
+
+        List<String> expIndexNames = Arrays.asList(
+            "foo_a_asc_idx",
+            "foo_b_asc_idx",
+            "foo_c_asc_idx",
+            "idx",
+            "IDX",
+            "iDx",
+            "",
+            " "
+        );
+
+        Collections.sort(expIndexNames);
+
+        checkIndexes(expIndexNames);
+
+        // Check that repeated queries fail.
+        for (String sql : createIndexQueries) {
+            GridTestUtils.assertThrows(
+                log,
+                () -> {
+                    runSql(sql);
+                    return null;
+                },
+                useJdbc ? SQLException.class : IgniteSQLException.class,
```

Comment:
Should we add error code check here?
E.g. IgniteQueryErrorCode.INDEX_ALREADY_EXISTS

Reply:
Agree, fixed.